gprofiler/profilers/java.py

```diff
@@ -110,6 +111,23 @@ def get_ap_log(self) -> str:
         return self._ap_log
 
 
+class AsyncProfiledProcessMonitor:
+    def __init__(self):
+        self._attached_processes = []
+        proc_events.register_exit_callback(self._proc_exit_callback)
+
+    def _proc_exit_callback(self, tid, pid, exit_code):
+        if pid in self._attached_processes:
+            if exit_code != 0:
+                logger.warning(
+                    f"Async-profiled Java process [{pid}] (TID: [{tid}]) exited with error code: {exit_code}"
+                )
+            self._attached_processes.remove(pid)
```

Comment:
This is not correct. A thread may exit and you should not remove the attached process upon it. I'm not sure of the exact semantics, but to my understandings only certain signals lead to killing the entire process.

I suggest that we focus only on the `tid == pid` case (== main thread) and figure out why we get `0` upon SIGKILL.. But first, let's move the proc_events code to utils please.

Reply:
Maybe the correct behavior should be to check whether `is_process_running` whenever the callback is called (it's invoked for each thread) and remove it only if it's not?

Reply:
Decided to check on TID == PID